cpp/src/arrow/compute/cast.cc

```diff
@@ -226,19 +254,219 @@ struct CastFunctor<O, I,
 };
 
 // ----------------------------------------------------------------------
+// Dictionary to other things
+
+template <typename IndexType>
+void UnpackFixedSizeBinaryDictionary(FunctionContext* ctx, const Array& indices,
+                                     const FixedSizeBinaryArray& dictionary,
+                                     ArrayData* output) {
+  using index_c_type = typename IndexType::c_type;
+  const uint8_t* valid_bits = indices.null_bitmap_data();
+  INIT_BITSET(valid_bits, indices.offset());
+
+  const index_c_type* in =
+      reinterpret_cast<const index_c_type*>(indices.data()->buffers[1]->data()) +
+      indices.offset();
+  uint8_t* out = output->buffers[1]->mutable_data();
+  int32_t byte_width =
+      static_cast<const FixedSizeBinaryType&>(*output->type).byte_width();
+  for (int64_t i = 0; i < indices.length(); ++i) {
+    if (bitset_valid_bits & (1 << bit_offset_valid_bits)) {
+      const uint8_t* value = dictionary.Value(in[i]);
+      memcpy(out + i * byte_width, value, byte_width);
+    }
+    READ_NEXT_BITSET(valid_bits);
+  }
+}
+
+template <typename T>
+struct CastFunctor<
+    T, DictionaryType,
+    typename std::enable_if<std::is_base_of<FixedSizeBinaryType, T>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options, const Array& input,
+                  ArrayData* output) {
+    const DictionaryArray& dict_array = static_cast<const DictionaryArray&>(input);
+    const DictionaryType& type = static_cast<const DictionaryType&>(*input.type());
+    const DataType& values_type = *type.dictionary()->type();
+    const FixedSizeBinaryArray& dictionary =
+        static_cast<const FixedSizeBinaryArray&>(*type.dictionary());
+
+    // Check if values and output type match
+    DCHECK(values_type.Equals(*output->type))
```

Comment:
I think we still need the status here as the former checks only check for matching type ids, here we also need to check for the `byte_width`.

Reply:
Wouldn't this indicate that the user of this function made a programming error and created the wrong kind of output type?

Reply:
Yes and as this would probably bubble up in Python & co, I really would like to return Statuses instead of Segmentation faults.

Reply:
When I say "user of this function" I mean the developers inside e.g. the parquet-cpp implementation. I don't think that non-developer users should be exposed to a non-exceptional failure inside this function

Reply:
So in application code (e.g. in parquet-cpp), this type check (if it could possibly bubble up to users) should happen before invoking the cast kernel

Reply:
Looking at the use in https://github.com/apache/parquet-cpp/pull/393, this check would fail *only if* `AllocateIfNotPreallocated` set the incorrect type in `output`, which is something that would be caught by test coverage in Arrow and should not be possible to bubble up to say a Python user or another third party user of parquet-cpp. 

If a third party developer were to pass a pre-allocated `FixedSizeBinaryArray`, then they are responsible for setting `output->type` to the correct value before invoking this function; if we provide any guarantees of runtime type checkout of such arguments, it will be a slippery slope and cause a lot of code bloat. In our internal use of the kernels, the type safety should be enforced as high in the call stack as possible, and similarly third party users of these functions must accept the responsibility of passing the right data structure if they are doing something advanced (like writing into preallocated memory)